### PR TITLE
replace the celery-based task queue with a kombu-based implementation

### PR DESCRIPTION
**Pull request by ryanpetrello**
This change implements `awx-manage run_dispatcher`, a new command that binds to RabbitMQ via kombu and balances messages across a pool of worker processes that are similar to celeryd workers in spirit.  Specifically, this includes:

- a new decorator, `awx.main.dispatch.task`, which can be used to decorate functions or classes so that they can be designated as "Tasks"
- support for fanout/broadcast tasks (at this point in time, only `conf.Setting` memcached flushes use this functionality)
- support for job reaping
- support for success/failure hooks for job runs (i.e., `handle_work_success` and `handle_work_failure`)
- support for auto scaling worker pool that scale processes up and down on demand
- minimal support for RPC, such as status checks and pool recycle/reload

### CODE REVIEW

**Comment by ryanpetrello** [comment]
I'm still testing this extensively, but given the scope of this change, I think it's time for folks to start taking a peek at the contents of this change and offer feedback.

File: Makefile [hunk 1]
```diff
@@ -250,7 +250,7 @@ server_noattach:
 	tmux new-session -d -s awx 'exec make uwsgi'
 	tmux rename-window 'AWX'
 	tmux select-window -t awx:0
-	tmux split-window -v 'exec make celeryd'
+	tmux split-window -v 'exec make dispatch'
```
Comment (ryanpetrello) [inline Makefile]: Typo.

File: Makefile [hunk 2]
```diff
@@ -298,7 +292,7 @@ uwsgi: collectstatic
 	@if [ "$(VENV_BASE)" ]; then \
 		. $(VENV_BASE)/awx/bin/activate; \
 	fi; \
-    uwsgi -b 32768 --socket 127.0.0.1:8050 --module=awx.wsgi:application --home=/venv/awx --chdir=/awx_devel/ --vacuum --processes=5 --harakiri=120 --master --no-orphans --py-autoreload 1 --max-requests=1000 --stats /tmp/stats.socket --lazy-apps --logformat "%(addr) %(method) %(uri) - %(proto) %(status)" --hook-accepting1-once="exec:/bin/sh -c '[ -f /tmp/celery_pid ] && kill -1 `cat /tmp/celery_pid` || true'"
+    uwsgi -b 32768 --socket 127.0.0.1:8050 --module=awx.wsgi:application --home=/venv/awx --chdir=/awx_devel/ --vacuum --processes=5 --harakiri=120 --master --no-orphans --py-autoreload 1 --max-requests=1000 --stats /tmp/stats.socket --lazy-apps --logformat "%(addr) %(method) %(uri) - %(proto) %(status)" --hook-accepting1-once="exec:awx-manage run_dispatcher --reload"
```
Comment (ryanpetrello) [inline Makefile:278]: When uwsgi in dev reloads due to code changes, it will also call `awx-manage run_dispatcher --reload`; this will cause the pool to kill and create new workers, thus reloading the code in the dispatcher processes.

File: awx/main/dispatch/control.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import logging
```
Comment (ryanpetrello) [inline awx/main/dispatch/control.py:1]: This module contains code used to send RPC commands to the dispatcher process.  On my first pass, I've implemented three:

`awx-manage run_dispatcher --status`; this sends a message to the dispatcher that causes it to gather status information on the worker pool (how many processes, what they're working on) and report that data *back* to the user over AMQP.

`awx-manage run_dispatcher --running`; this sends a message to the dispatcher that causes it to reply with a list of task UUIDs it's currently running or has queued to run next.

`awx-manage run_dispatcher --reload` makes the worker process spin down and replace all of its workers; this PR uses that in the dev environment to accomplish automatic code as Python code changes.

File: awx/main/dispatch/control.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        )
+        producer.publish(msg, expiration=5, **kwargs)
+
+    def status(self, host=None, timeout=5):
+        host = host or settings.CLUSTER_HOST_ID
+        logger.warn('checking {} status for {}'.format(self.service, host))
+        reply_queue = Queue(name="amq.rabbitmq.reply-to")
+        with Connection(settings.BROKER_URL) as conn:
+            with Consumer(conn, reply_queue, callbacks=[self.process_message], no_ack=True):
+                self.publish({'control': 'status'}, conn, host, reply_to='amq.rabbitmq.reply-to')
```
Comment (ryanpetrello) [inline awx/main/dispatch/control.py]: If you're unfamiliar w/ `amq.rabbitmq.reply-to`, this is a good read that explains what's going on here:

https://medium.com/@nqbao/direct-reply-to-in-rabbitmq-76d070f55db0

File: awx/main/dispatch/pool.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
+import logging
```
Comment (ryanpetrello) [inline awx/main/dispatch/pool.py:1]: The code in this module implements the worker pool, and it's code that's shared between the new dispatcher *and* the callback receiver; for the *dispatcher*, it includes autoscaling code that forks new worker processes, checks their health and replaces ones that exit/die, and spins down workers that are idle.

```python
    class MessagePrinter(awx.main.dispatch.worker.BaseWorker):

         def perform_work(self, body):
             print body

     pool = WorkerPool(min_workers=4)  # spawn four worker processes
     pool.init_workers(MessagePrint().work_loop)
     pool.write(
         0,  # preferred worker 0
         'Hello, World!'
     )
```

As AMQP messages for the queues we care about come in, we call `pool.write()`, and the messages are distributed amongst the worker processes in the pool.  There are two different *worker* implementations - one for handling callback events, and one for running tasks (what celery *used* to do).

File: awx/main/dispatch/publish.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
+import inspect
```
Comment (ryanpetrello) [inline awx/main/dispatch/publish.py:1]: This code implements the `@task()` decorator.  You can decorate any function or class (you'll see the associated changes in `awx.main.tasks` in this diff), and the decorator will make that class/function have an `apply_async()` and `delay()` function that match celery's interface.  In this way, most of the code that used to use celery for queueing a task hasn't changed much (this is backwards compatible with how you use celery).

File: awx/main/dispatch/reaper.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+from django.db.models import Q
+from django.utils.timezone import now as tz_now
+from django.contrib.contenttypes.models import ContentType
+
+from awx.main.models import Instance, UnifiedJob, WorkflowJob
+
+logger = logging.getLogger('awx.main.dispatch')
+
+
+def reap(instance=None, status='failed'):
```
Comment (ryanpetrello) [inline awx/main/dispatch/reaper.py:29]: I extracted the job reaping code into a distinct function that we can test here.  It is *no longer* called in the periodic task manager.  Instead, it's called in a few strategic locations:

1.  When a dispatcher exits gracefully (i.e., via `SIGTERM`) and when it starts, it reaps any running jobs that match *its* execution node.  In this way, if the dispatcher restarts (or the node reboots), it cleans up after itself when it stops and when it comes back online:
https://github.com/ansible/awx/pull/2266/files#diff-f5abf60ba3722d35bf31ee73157af353R96
https://github.com/ansible/awx/pull/2266/files#diff-9d4ea1dd908b35fb92eaede4bd10bb46R124
2.  When the dispatcher notices that a worker processes has exited with an unacknowledged task it was working on (it never placed the task in its "finished" queue signifying that it was done), that job is automatically marked as failed by the reaper: https://github.com/ansible/awx/pull/2266/files#diff-0ce27c4edcd6336eddfce36efed72a0fR333
3.  If a node exits suddenly (i.e., `SIGKILL`, power loss) and it _never_ comes back online, other nodes in a cluster will notice and set the capacity to zero.  With the changes in _this_ PR, that process *also* calls this reaping function to mark jobs as failed for nodes that go missing in a cluster:
https://github.com/ansible/awx/pull/2266/files#diff-9d4ea1dd908b35fb92eaede4bd10bb46R362

This implementation actually improves on one of the limitations in the 3.3.0 implementation - namely that on a single node install, if celery node goes offline during a playbook run, there's no periodic task manager to discover and reap jobs stuck in `Running`.

File: awx/main/dispatch/worker/base.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
+# Copyright (c) 2018 Ansible by Red Hat
```
Comment (ryanpetrello) [inline awx/main/dispatch/worker/base.py]: This module contains the main entry point for the dispatcher...

File: awx/main/dispatch/worker/base.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        for i, worker in enumerate(consumer.pool.workers):
+            if len(worker.managed_tasks) < size:
+                queue = i
+                size = len(worker.managed_tasks)
+        return queue
+
+
+class AWXConsumer(ConsumerMixin):
+
+    def __init__(self, name, connection, worker, queues=[], pool=None, strategy=RoundRobinStrategy):
```
Comment (ryanpetrello) [inline awx/main/dispatch/worker/base.py]: ...much like the callback receiver implementation in 3.3.0 (on which this code is based), this entry point is a `kombu.ConsumerMixin`.  This mixin lets us define a set of queues we care about (the global default queue, the instance's hostname-specific queue, and the broadcast queue), fetch new events from those queues, and write them to the pool of workers using what I've called a _strategy_).

There are two strategies defined in this module:

1.  `RoundRobinStrategy` matches the behavior of the 3.3.0 callback receiver.  The goal is to pass events to the pool of workers as *fast* as possible using `UUID modulo total_workers`.  We don't care much about even distribution, because the handling of each task (saving an event to the database) is super fast.

2.  `LeastBusyStrategy` - this is the strategy used for task execution.  The idea is that we should give new tasks to the process that is the *least* busy - in this way, we aim to ensure that each worker is as busy as possible (we don't want Worker A to have zero tasks and Worker B to have a queue with 10 tasks sitting in it, because a long running task could hold things up).

File: awx/main/dispatch/worker/base.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        self.worker.on_stop()
+        raise SystemExit()
+
+
+class BaseWorker(object):
+
+    def work_loop(self, queue, finished, idx, *args):
+        ppid = os.getppid()
+        signal_handler = WorkerSignalHandler()
+        while not signal_handler.kill_now:
```
Comment (ryanpetrello) [inline awx/main/dispatch/worker/base.py:116]: This is the main worker loop.  Every child process that we fork and add to the worker pool works this way:

1.  While `not kill_now`:
2.  Wait up to one second for a new message from my queue and call `perform_work()` on that message (if any).
3.  `def perform_work` is defined in subclasses in the two worker implementations.  One of those implementations handles callback events (it's almost entirely the code from the 3.3.0 callback receiver).  The other implementation is the one that does what celery did in 3.3.0 - takes tasks, deserializes them, and runs Python code.

File: awx/main/dispatch/worker/base.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+            except Exception as e:
+                logger.error("Exception on worker, restarting: " + str(e))
+                continue
+            try:
+                self.perform_work(body, *args)
+            finally:
+                if 'uuid' in body:
+                    uuid = body['uuid']
+                    logger.debug('task {} is finished'.format(uuid))
+                    finished.put(uuid)
```
Comment (ryanpetrello) [inline awx/main/dispatch/worker/base.py:136]: When a task has been finished, we put it into a `multiprocessing.Queue` so the pool knows we finished it and can track state.

File: awx/main/dispatch/worker/callback.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+from awx.main.consumers import emit_channel_notification
+from awx.main.models import (JobEvent, AdHocCommandEvent, ProjectUpdateEvent,
+                             InventoryUpdateEvent, SystemJobEvent, UnifiedJob)
+
+from .base import BaseWorker
+
+logger = logging.getLogger('awx.main.dispatch')
+
+
+class CallbackBrokerWorker(BaseWorker):
```
Comment (ryanpetrello) [inline awx/main/dispatch/worker/callback.py:20]: This is the worker implementation for callback events.  This is almost 100% copied from the callback receiver command code in 3.3.0; not much has changed here outside of the refactoring described above.

File: awx/main/dispatch/worker/task.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+import six
+
+from awx.main.tasks import dispatch_startup, inform_cluster_of_shutdown
+
+from .base import BaseWorker
+
+logger = logging.getLogger('awx.main.dispatch')
+
+
+class TaskWorker(BaseWorker):
```
Comment (ryanpetrello) [inline awx/main/dispatch/worker/task.py:16]: This worker implementation takes messages published via `.apply_async()`, deserializes them, and runs native Python code.  This does what celery workers did in 3.3.0.

File: awx/main/management/commands/run_dispatcher.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    help = 'Launch the task dispatcher'
+
+    def add_arguments(self, parser):
+        parser.add_argument('--status', dest='status', action='store_true',
+                            help='print the internal state of any running dispatchers')
+        parser.add_argument('--reload', dest='reload', action='store_true',
+                            help=('cause the dispatcher to recycle all of its worker processes;'
+                                  'running jobs will run to completion first'))
+
+    def beat(self):
```
Comment (ryanpetrello) [inline awx/main/management/commands/run_dispatcher.py:37]: We don't use celery workers anymore, but we're still using celerybeat for now for periodic tasks, because it works.  We _may_ replace this in the future if we really want to lose the dependency on the celery library.

I don't intend to change this in the _near_ future.

File: awx/main/management/commands/run_dispatcher.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        beat = Process(target=self.beat)
+        beat.daemon = True
+        beat.start()
+
+        reaper.reap()
+        consumer = None
+        with Connection(settings.BROKER_URL) as conn:
+            try:
+                bcast = 'tower_broadcast_all'
+                queues = [
```
Comment (ryanpetrello) [inline awx/main/management/commands/run_dispatcher.py:101]: No more dynamic queue binding wackiness via celery signals.  At startup time, we very explicitly decide which queues to listen on.

File: awx/main/models/ha.py [hunk 16]
```diff
@@ -32,7 +32,7 @@
 
 
 def validate_queuename(v):
-    # celery and kombu don't play nice with unicode in queue names
+    # kombu doesn't play nice with unicode in queue names
```
Comment (ryanpetrello) [inline awx/main/models/ha.py:35]: This probably isn't really true anymore; at a later date, we could probably investigate lifting this restriction on unicode instance group names.

File: awx/main/scheduler/task_manager.py [hunk 17]
```diff
@@ -85,79 +74,6 @@ def get_tasks(self, status_list=('pending', 'waiting', 'running')):
                            key=lambda task: task.created)
         return all_tasks
 
-    '''
```
Comment (ryanpetrello) [inline awx/main/scheduler/task_manager.py:88]: All of the job reaping code has been removed from the task manager.

File: awx/main/tasks.py [hunk 18]
```diff
@@ -87,79 +82,40 @@
...
-        else:
-            logger.exception(six.text_type('Task {} encountered exception.').format(self.name), exc_info=exc)
-    except Exception:
-        # It's fairly critical that this code _not_ raise exceptions on logging
-        # If you configure external logging in a way that _it_ fails, there's
-        # not a lot we can do here; sys.stderr.write is a final hail mary
-        _, _, tb = sys.exc_info()
-        traceback.print_tb(tb)
-
+def dispatch_startup():
```
Comment (ryanpetrello) [inline awx/main/tasks.py:85]: This function gets called when the dispatcher starts.  It does all the stuff we used to do in the celeryd init signal.

File: awx/main/tasks.py [hunk 19]
```diff
@@ -413,6 +367,10 @@ def cluster_node_heartbeat(self):
                 other_inst.save(update_fields=['capacity'])
                 logger.error(six.text_type("Host {} last checked in at {}, marked as lost.").format(
                     other_inst.hostname, other_inst.modified))
+                try:
```
Comment (ryanpetrello) [inline awx/main/tasks.py]: If a node's capacity is set to zero by one of its peers, reap its jobs, too.

File: awx/main/tasks.py [hunk 20]
```diff
@@ -629,7 +586,7 @@ def delete_inventory(self, inventory_id, user_id):
             return
         except DatabaseError:
             logger.exception('Database error deleting inventory {}, but will retry.'.format(inventory_id))
-            self.retry(countdown=10)
+            # TODO: self.retry(countdown=10)
```
Comment (ryanpetrello) [inline awx/main/tasks.py:594]: @AlanCoding haven't figured this one out yet :/

I may just put this entire function in a loop and do retries with exception handling without spawning new tasks.  I don't really want to implement celery-style retries just for this.

File: awx/main/tasks.py [hunk 21]
```diff
@@ -1160,7 +1114,7 @@ def build_passwords(self, job, **kwargs):
         Build a dictionary of passwords for SSH private key, SSH user, sudo/su
         and ansible-vault.
         '''
-        passwords = super(RunJob, self).build_passwords(job, **kwargs)
+        passwords = BaseTask.build_passwords(self, job, **kwargs)
```
Comment (ryanpetrello) [inline awx/main/tasks.py]: Had to change this (here and below) because of the way I implemented the `@task` decorator; making it work the same w/ `super()` required some pretty gross metaclassing nastiness that wasn't easy to understand (it turns out celery does something similar), so I just went with this.

File: awx/main/utils/autoscale.py [hunk 22]
```diff
@@ -1,27 +0,0 @@
...
-
-
-class DynamicAutoScaler(Autoscaler):
-
-    def __init__(self, pool, max_concurrency, min_concurrency=0, worker=None,
-                 keepalive=AUTOSCALE_KEEPALIVE, mutex=None):
-        super(DynamicAutoScaler, self).__init__(pool, max_concurrency,
-                                                min_concurrency, worker,
-                                                keepalive, mutex)
-        settings_absmem = getattr(settings, 'SYSTEM_TASK_ABS_MEM', None)
```
Comment (ryanpetrello) [inline awx/main/utils/autoscale.py:16]: This code found another home in `awx.main.dispatcher.pool`'s autoscaling code.

File: awx/main/utils/reload.py [hunk 23]
```diff
@@ -48,6 +47,6 @@ def _supervisor_service_command(service_internal_names, command, communicate=Tru
         logger.info('Submitted supervisorctl {} command, not waiting for result'.format(command))
 
 
-def stop_local_services(service_internal_names, communicate=True):
-    logger.warn('Stopping services {} on this node in response to user action'.format(service_internal_names))
-    _supervisor_service_command(service_internal_names, command='stop', communicate=communicate)
+def stop_local_services(communicate=True):
```
Comment (ryanpetrello) [inline awx/main/utils/reload.py:42]: It turns out we don't actually call `stop_local_services` with explicit service names anymore - the only place we use this is for capacity=0 due to version mismatch where we stop _everything_, so I simplified it a bit.

@AlanCoding with this change, there's honestly no longer any need to track `settings.SERVICE_NAME_DICT`.

File: awx/settings/defaults.py [hunk 24]
```diff
@@ -459,40 +457,9 @@ def IS_TESTING(argv=None):
 # Set default ports for live server tests.
 os.environ.setdefault('DJANGO_LIVE_TEST_SERVER_ADDRESS', 'localhost:9013-9199')
 
-djcelery.setup_loader()
-
 BROKER_POOL_LIMIT = None
 BROKER_URL = 'amqp://guest:guest@localhost:5672//'
-CELERY_EVENT_QUEUE_TTL = 5
 CELERY_DEFAULT_QUEUE = 'awx_private_queue'
```
Comment (ryanpetrello) [inline awx/settings/defaults.py:464]: I kept a few of these setting names w/ `CELERY_` because it was just simpler, and I didn't want to muck with setting names, given their historic fragility and the fact that people may have old setting files laying around w/ custom values.

File: installer/roles/image_build/files/supervisor_task.conf [hunk 25]
```diff
@@ -50,7 +38,7 @@ stderr_logfile=/dev/stderr
 stderr_logfile_maxbytes=0
 
 [group:tower-processes]
-programs=celery,celery-watcher,callback-receiver,channels-worker
+programs=dispatcher,callback-receiver,channels-worker
```
Comment (ryanpetrello) [inline installer/roles/image_build/files/supervisor_task.conf:41]: @shanemcd we'll have to copy this into wherever it belongs for prod OpenShift builds (is there a similar OpenShift-oriented version of this?)

File: awx/main/tasks.py (hunk 20 above)
Comment (AlanCoding) [reply awx/main/tasks.py:594]: I think that would be fine. There's only a certain subset of errors that the retry addressed - specifically I think this was due to deadlocks with other processes, which is inherently transient. That should still be covered by a sleep and try again approach.

File: awx/main/utils/reload.py (hunk 23 above)
Comment (AlanCoding) [reply awx/main/utils/reload.py:42]: The code where we called this probably never worked with OpenShift to be perfectly honest... plus, that doesn't get called in normal operating procedure, it's only if things have gone sideways. With this celery replacement, we might eventually develop some other method of standing still and not doing anything until a problem is fixed.

**Comment by wenottingham** [comment]
Can you test this with current kombu? We might as well update while we're here (and we have other reasons to...)

**Comment by ryanpetrello** [comment]
@wenottingham that's a great point - once I've got integration reliably passing, I'll give that a shot, too.  I'm trying to compare apples to apples on a first pass here.

File: awx/main/tasks.py [hunk 26]
```diff
@@ -87,91 +82,56 @@
...
-        queues.append(Broadcast(q.encode('utf-8')))
-    conf.CELERY_QUEUES = list(set(queues))
-
-    # Expedite the first hearbeat run so a node comes online quickly.
-    cluster_node_heartbeat.apply([])
 
-
-@worker_shutdown.connect
-def inform_cluster_of_shutdown(*args, **kwargs):
+def inform_cluster_of_shutdown():
```
Comment (ryanpetrello) [inline awx/main/tasks.py:118]: This function gets called when the dispatcher stops.  It sets capacity to zero and reaps any running jobs assigned to the node.

File: awx/main/tasks.py (hunk 21 above)
Comment (AlanCoding) [reply awx/main/tasks.py]: Could you leave a comment (or many comments) to this effect? I see this as trivial to work around from a developer perspective, but there are no guardrails to prevent someone from doing this, which would otherwise be ingrained in muscle memory and best practice.

File: docs/tasks.md [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+-------------------------
+
+AWX publishes tasks in two distinct ways.
+
+*Direct* messages are bound _directly_ to a specific named queue.  When you launch
+a Job Template in AWX, it looks at the available capacity of the various nodes
+in your cluster and chooses an `Execution Node` where the playbook will run.
+In this scenario, AWX publishes a message to a direct queue associated with
+that AWX node.  The worker processes running on that AWX node are specifically
+bound to _listen_ for new events on their named queue.
```
Comment (AlanCoding) [inline docs/tasks.md]: The worker processes listen for new events on the queue named for the _instance_?

File: awx/main/dispatch/control.py (hunk 3 above)
Comment (AlanCoding) [reply awx/main/dispatch/control.py:1]: you said you implemented two - was the other one the command to recycle workers after jobs finished?

File: docs/tasks.md (hunk 27 above)
Comment (ryanpetrello) [reply docs/tasks.md]: Yea, that's probably a clearer way to describe that.

File: awx/main/dispatch/control.py (hunk 3 above)
Comment (ryanpetrello) [reply awx/main/dispatch/control.py:1]: (updated my comment)

File: awx/main/dispatch/control.py [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+        )
+        producer.publish(msg, expiration=5, **kwargs)
+
+    def status(self, *args, **kwargs):
+        return self.control_with_reply('status', *args, **kwargs)
+
+    def running(self, *args, **kwargs):
+        return self.control_with_reply('running', *args, **kwargs)
+
+    def control_with_reply(self, command, host=None, timeout=5):
```
Comment (ryanpetrello) [inline awx/main/dispatch/control.py:38]: If you're unfamiliar with `amq.rabbitmq.reply-to`, here's a great article explaining what's going on here:

https://medium.com/@nqbao/direct-reply-to-in-rabbitmq-76d070f55db0

File: awx/main/dispatch/worker/base.py [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+
+    def __init__(self):
+        self.kill_now = False
+        signal.signal(signal.SIGINT, self.exit_gracefully)
+
+    def exit_gracefully(self, *args, **kwargs):
+        self.kill_now = True
+
+
+class AWXConsumer(ConsumerMixin):
```
Comment (ryanpetrello) [inline awx/main/dispatch/worker/base.py:35]: The main entry point for the callback receiver and dispatcher is a `kombu.ConsumerMixin`.  This mixin allows you to define `get_consumers()` and listen for events on one or more RabbitMQ queues.  As new messages come one, they're passed to a callback function (`self.process_task`) which hands them off to a worker in our pool.  This is the same pattern we use to process callback events in 3.3.0 and prior.

Additionally, this class contains special logic to capture control messages (`--status`, `--reload`, `--running`) and reply to the client (see `awx.main.dispatch.control`).

File: awx/main/dispatch/worker/base.py [hunk 30]
```diff
@@ -0,0 +1,146 @@
...
+                if 'uuid' in body:
+                    uuid = body['uuid']
+                    logger.debug('task {} is finished'.format(uuid))
+                    finished.put(uuid)
+        logger.warn('worker exiting gracefully pid:{}'.format(os.getpid()))
+
+    def perform_work(self, body):
+        raise NotImplementedError()
+
+    def on_start(self):
```
Comment (ryanpetrello) [inline awx/main/dispatch/worker/base.py:142]: These are hooks that can be defined in the subclasses to do things *before* and *after* we start consuming AMQP messages.  They're mostly used by the task worker to:

- Perform ancillary functions at startup, such as instance auto-registration in OpenShift, and initial heartbeat (`awx.main.tasks.dispatch_startup`).
- Set capacity to zero and reap running jobs at shutdown.

File: awx/main/tasks.py [hunk 31]
```diff
@@ -1109,12 +1067,12 @@ def get_ssh_key_path(self, instance, **kwargs):
         return ''
 
 
+@task
 class RunJob(BaseTask):
     '''
     Celery task to run a job using ansible-playbook.
     '''
 
-    name = 'awx.main.tasks.run_job'
```
Comment (ryanpetrello) [inline awx/main/tasks.py:1117]: Giving names to class-based tasks is no longer necessary.

File: awx/main/tasks.py [hunk 32]
```diff
@@ -280,20 +240,18 @@ def apply_cluster_membership_policies(self):
...
 
-@shared_task(exchange='tower_broadcast_all', bind=True)
-def handle_setting_changes(self, setting_keys):
+@task(queue='tower_broadcast_all', exchange_type='fanout')
+def handle_setting_changes(setting_keys):
     orig_len = len(setting_keys)
     for i in range(orig_len):
         for dependent_key in settings_registry.get_dependent_settings(setting_keys[i]):
             setting_keys.append(dependent_key)
-    logger.warn('Processing cache changes, task args: {0.args!r} kwargs: {0.kwargs!r}'.format(
```
Comment (ryanpetrello) [inline awx/main/tasks.py:289]: I got rid of this because it was noisy.

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

**Comment by Spredzy** [comment]
recheck

File: Makefile (hunk 2 above)
Comment (chrismeyersfsu) [reply Makefile:278]: This makes me happy.

File: awx/__init__.py [hunk 33]
```diff
@@ -12,14 +12,6 @@
 __all__ = ['__version__']
 
 
-# Isolated nodes do not have celery installed
```
Comment (chrismeyersfsu) [inline awx/__init__.py:15]: :+1: hack removed

File: awx/main/dispatch/__init__.py [hunk 34]
```diff
@@ -0,0 +1,5 @@
+from awx.main.models import Instance
+
+
+def get_local_queuename():
+    return Instance.objects.me().hostname.encode('utf-8')
```
Comment (chrismeyersfsu) [inline awx/main/dispatch/__init__.py]: Might want to stick with settings.CLUSTER_HOST_ID. I'm thinking of awx in openshift "mode" where offline nodes will be deleted by other nodes. If a node is deleted from the DB, but is still running jobs (creating events) and processing events from rabbitmq to the DB; getting it's hostname will fail because it will be deleted from the DB. 

Arguably, this node was decided dead by the other nodes and thus the node should not be doing work anymore. However, awx kind of has an on-the-fence behavior when it's offline. We don't hard kill jobs so those keep running and it's not deterministic as to if events created from that job run are processed or not.

TL;DR I want to work towards being able to concisely describe awx behavior when a node goes offline.

File: awx/main/dispatch/reaper.py (hunk 7 above)
Comment (chrismeyersfsu) [reply awx/main/dispatch/reaper.py:29]: *If a celery node goes offline gracefully during a playbook run; the new implementation will handle this. This implementation doesn't solve the super sharp edge case of who watches the watcher. i.e. case 3 w/ a single node install will result in jobs stuck in `running`.

File: awx/main/dispatch/reaper.py [hunk 35]
```diff
@@ -0,0 +1,46 @@
...
+    '''
+    Reap all jobs in waiting|running for this instance.
+    '''
+    me = instance or Instance.objects.me()
+    now = tz_now()
+    workflow_ctype_id = ContentType.objects.get_for_model(WorkflowJob).id
+    jobs = UnifiedJob.objects.filter(
+        (
+            Q(status='running') |
+            Q(status='waiting', modified__lte=now - timedelta(seconds=60))
```
Comment (chrismeyersfsu) [inline awx/main/dispatch/reaper.py:39]: This problem remains unsolved. The task manager gaits ansible jobs from running based on capacity that it knows the underlying host to have. Disjointedly, the celery/dispatcher spins up and down workers based on capacity.

The task manager decides to transition a job from `pending` to `waiting` when (1) there is no business logic blocking it and (2) there is capacity. BUT there could _actually_ not be enough capacity (dispatch workers) to pick the job up. We choose to allow 60 seconds for someone to pick the job up before the job is a candidate for killing it.

To actually solve this problem, the task manager and dispatcher would need to talk to each other to ensure that the there is truly capacity. Or, the two systems need to make their decisions based on the same inputs (i.e. memory and capacity).

File: awx/main/dispatch/__init__.py (hunk 34 above)
Comment (ryanpetrello) [reply awx/main/dispatch/__init__.py]: Don't we enforce `Instance.objects.me().hostname == settings.CLUSTER_HOST_ID` now?

File: awx/main/dispatch/reaper.py (hunk 7 above)
Comment (ryanpetrello) [reply awx/main/dispatch/reaper.py:29]: Yea.  If there's a single node and the node goes down, there _is_ no watcher to clean up "Running" jobs.  But that was the case before with the task manager right?  The task manager will never run if there's no celery/dispatcher running anywhere.

File: awx/main/dispatch/reaper.py (hunk 35 above)
Comment (ryanpetrello) [reply awx/main/dispatch/reaper.py:39]: Agreed.  Longer term, I'd really like to have the task manager and dispatcher communicate closer to realtime - there are cases where the heartbeat isn't fast enough, and could result in jobs being sent to nodes that are actually down.  Maybe the RPC bits that I have in this PR could be used in the future to determine real-time capacity?  For example maybe before the task scheduler "picks" a host, it sends an RPC message (w/ a few-second timeout) to determine its *current* capacity/availability.

Comment (chrismeyersfsu) [reply awx/main/dispatch/reaper.py:39]: Because the task manager runs on any node and can decide the run a task on another node; not sure that's going to be possible.

File: awx/main/dispatch/__init__.py (hunk 34 above)
Comment (chrismeyersfsu) [reply awx/main/dispatch/__init__.py]: Yep, but `Instance.objects.me()` relies on a query to the database to find itself. Itself can be deleted if it's deemed to be offline.

**Comment by ryanpetrello** [comment]
recheck

File: awx/main/dispatch/__init__.py (hunk 34 above)
Comment (ryanpetrello) [reply awx/main/dispatch/__init__.py]: Okay, that's fair.  I'll update this to use `settings.CLUSTER_HOST_ID`.

**Review by matburt — APPROVED** [review]

**Comment by matburt** [comment]
This all looks good to me.

**Comment by Spredzy** [comment]
recheck

**Comment by matburt** [comment]
recheck

File: awx/main/dispatch/reaper.py (hunk 7 above)
Comment (chrismeyersfsu) [reply awx/main/dispatch/reaper.py:29]: Correct, figured it was worth stating for completeness.

File: awx/main/dispatch/reaper.py (hunk 35 above)
Comment (ryanpetrello) [reply awx/main/dispatch/reaper.py:39]: It _should be_ possible for a task manager running code on Node A to communicate to the dispatcher on Node B via the AMQP RPC I've written here. That's an exercise for another PR, though (we probably don't even care about _actual_ capacity, just a "hey, before I give you a job to run, are you alive and responsive via AMQP?")

**Comment by ryanpetrello** [comment]
recheck

**Review by wwitzel3 — APPROVED** [review]
